File: sorting/Sort2.cpp
#include <iostream>
#include <vector>

using namespace std;

class BinaryTreeSort{
    public:
        
        // Merge sort, use array
        void mergeSort(int a[], int l, int r){
            // If l == r, do nothing
            if (l < r){
                int mid = (l+r) / 2; // Or use mid = l + (r-l)/2 for avoiding overflow
                mergeSort(a, l, mid);
                mergeSort(a, mid+1, r);
                merge(a, l, mid, r);
            }
        }
        // merge a[l, mid] and a[mid, r] of a[]
        void merge(int a[], int l, int mid, int r){
            int i, j, k;
            int n1 = mid - l + 1; // remember + 1 here
            int n2 = r - mid;
            // Create temp arrays
            int left[n1], right[n2];
            // Copy data to temp
            for (i=0; i < n1; i++)
                left[i] = a[l + i];
            for (j=0; j < n2; j++)
                right[j] = a[mid+1 + j];
            //Merge
            i = 0;
            j = 0;
            k = l; // Not 0 !!!
            while (i < n1 && j < n2){
                if (left[i] <= right[j]){
                    a[k] = left[i];
                    i++;
                }else{
                    a[k] = right[j];
                    j++;
                }
                k++;
            }
            // Copy the remaining if any
            while (i < n1){
                a[k] = left[i];
                k++;
                i++;
            }
            while (j < n2){
                a[k] = right[j];
                j++;
                k++;
            }
        }

        void printArray(int a[], int sz) { 
            for (int i = 0; i < sz; i++)
                cout << a[i] << " "; 
            cout << endl; 
        }
};

int main(){
    BinaryTreeSort bts;
    int a[] = {21, 2, 45, 103, 0, 64, 0};
    int sz = sizeof(a)/sizeof(a[0]);
    
    bts.printArray(a, sz);
    bts.mergeSort(a, 0, sz-1);
    bts.printArray(a, sz);
    return 0;
}

File: sorting/sort2.py
# Merge Sort
# divide and conquer
# Time: O(n log n), O(n log n), O(n log n)
# Space: O(n)
def mergesort(aList):
    if len(aList) <= 1:
        return aList
    else:
        mid = len(aList)/2
        left = mergesort(aList[0:mid])
        right = mergesort(aList[mid:])
        # If two sub-list already sorted
        if left[-1] <= right[0]:
            return left + right
        else:
            return merge(left, right)


def merge(left, right):
    result = list()
    while len(left) > 0 and len(right) > 0:
        if left[0] <= right[0]:
            result.append(left[0])
            # This works even list has only one element left
            left = left[1:]
        else:
            result.append(right[0])
            right = right[1:]
    if len(left) > 0:
        result = result + left
    if len(right) > 0:
        result = result + right
    return result


# Quick sort
# O(n log n), divide-and-conquer, in-place, short inner loop
# Time: O(n log n), O(n log n), O(n^2)
# Space: O(log n)
"""This extra space is used in stack frames, it's not practical for making it
iterative because it's not tail recursive
"""
"""
When is QuickSort impractical?
Ans:
    1. Any situation when you do not have random access in O(1), i.e. Linked list
    2. Distributed sorting. i.e. sort data across files
"""
def quicksort(a, left, right):
    i = left
    j = right
    pivot = a[(left + right) / 2]

    while (i <= j):
        while (a[i] < pivot):
            i = i + 1
        while (a[j] > pivot):
            j = j - 1
        if (i <= j):
            # swap a[i], a[j]
            a[i], a[j] = a[j], a[i]
            i = i + 1
            j = j - 1

    if (left < j):
        quicksort(a, left, j)
    if (i < right):
        quicksort(a, i, right)


# Heapsort
# O(n·lg(n)) in-place sort, but is not stable.
# Heapsort is not stable because operations on the heap can change the relative order of equal items.
# Time: O(n log n), O(n log n), O(n log n)
# Space: O(1)
#* time complexity of building a heap is: O(n)
def heapsort(a):
    def swap(a, i, j):
       a[i], a[j] = a[j], a[i]

    def siftdown(a, i, size):
        left = 2*i +1
        right = 2*i +2
        largest = i

        if left < size-1 and a[left] > a[i]:
            largest = left
        if right < size-1 and a[right] > a[largest]:
            largest = right
        if largest != i:
            swap(a, i, largest)
            siftdown(a, largest, size)

    # a bottom-up manner
    def heapify(a, size):
        p = size/2 - 1
        while p >= 0:
            siftdown(a, p, size)
            p -= 1

    size = len(a)
    heapify(a, size)
    end = size - 1
    while end > 0:
        swap(a, 0, end)
        siftdown(a, 0, end)
        end -= 1

# Radix sort

aList = [21, 2, 45, 103, 0, 64, 0]
print mergesort(aList)
aList = [21, 2, 45, 103, 0, 64, 0]
quicksort(aList, 0, len(aList)-1)
print aList
aList = [21, 2, 45, 103, 0, 64, 0]
heapsort(aList)
print aList
